### PR TITLE
Created a single class business layer

### PR DESCRIPTION
**Pull request by paymog**
Before we go further into this, can you guys critique the fuck out of what I'm doing?

My next plan is to move some/most/all of the logic from CryptogramSolveWidget to CrytpogramManager.

I envision that each UI component will have some sort of xxxManager class to handle the logic it needs

### CODE REVIEW

File: pg13/business/CryptogramManager.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+	public void validatePlaintext(String string) {
+		char[] characters = string.toCharArray();
+		for (int i = 0; i < characters.length; i++)
+		{
+			char curr = characters[i];
+			if ("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890.,:;- !?()&#$%'\"".indexOf(curr) < 0)
+			{
+				throw new IllegalArgumentException("Invalid character in plaintext");
+			}
```
Comment (rikimbo) [inline pg13/business/CryptogramManager.java:41]: Should we really be throwing an exception for this?  This function is used by the GUI to validate text entered, and the user is going to try to type in all sorts of garbage.  Is it really an exceptional case?  Maybe it could be restructured to return a boolean.

EDIT:  Actually maybe it's legit to do it like that.  You've done a lot more GUI interaction then I have.  Is that what people do?

File: pg13/presentation/CryptogramEditWidget.java [hunk 2]
```diff
@@ -9,101 +9,73 @@
...
-				{  
-	            	case SWT.BS:           // Backspace  
-	            	case SWT.DEL:          // Delete  
-	            	case SWT.HOME:         // Home  
-	            	case SWT.END:          // End  
-	            	case SWT.ARROW_LEFT:   // Left arrow  
-	            	case SWT.ARROW_RIGHT:  // Right arrow  
-	                textIsValid = true;  
-				}  
-
```
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:82]: You shouldn't remove the part of the validation that checks the keycode for backspace, delete, and other control keys.  It might prevent the user from doing those things in the text box.  Not sure though.  Did you test it?

File: pg13/presentation/CryptogramEditWidget.java [hunk 3]
```diff
@@ -9,101 +9,73 @@
...
-				}
-				
-				// block the event if text is not valid
-				if (!textIsValid)
-				{
+		txtPlaintext.addVerifyListener(new VerifyListener() {
+			public void verifyText(VerifyEvent event) {
+				try {
+					cm.validatePlaintext(event.text);
+				} catch (IllegalArgumentException e) {
```
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:73]: I like the call to validate to a CM instead of doing this directly in here.  Still unsure about using an exception for it though.  See my other comment.

File: pg13/presentation/CryptogramEditWidget.java (hunk 2 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:82]: Yup, all of those work. I totally forgot to test for that though.

File: pg13/presentation/CryptogramEditWidget.java [hunk 4]
```diff
@@ -111,13 +83,11 @@ public void verifyText(VerifyEvent event)
 		fd_txtPlaintext.top = new FormAttachment(lblPlaintext, 6);
 		fd_txtPlaintext.left = new FormAttachment(cmpPreview, 0, SWT.LEFT);
 		txtPlaintext.setLayoutData(fd_txtPlaintext);
-		
+
 		Button btnPreview = new Button(this, SWT.NONE);
-		btnPreview.addSelectionListener(new SelectionAdapter() 
-		{
```
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:117]: Why did you change the bracketing?  I thought our standards said the bracketing should be on a separate line.

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:73]: I'm also unsure of whether an exception is the best way to handle it. I debated a boolean but my gut told me exception is the way to go. I'm not sure if I can defend my choice very well though.

File: pg13/presentation/CryptogramEditWidget.java [hunk 5]
```diff
@@ -131,20 +101,25 @@ public void widgetSelected(SelectionEvent e)
...
-		cmpPreview.displayCryptogram();
+	private void preview() {
+		try {
+			this.cm.setPlaintext(txtPlaintext.getText());
+			cmpPreview.setCryptogram(this.cm.getCryptogram());
+			cmpPreview.displayCryptogram();
+		} catch (IllegalArgumentException e) {
+			MessageBox dialog = new MessageBox(this.getShell(), SWT.OK);
+			dialog.setText("Invalid text");
+			dialog.setMessage("The text you have entered is invalid. Please modify the text.");
```
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:117]: This shouldn't ever happen, since everything that's entered into the plaintext box is intercepted by a listenner there, but still, it's good to have it.

EDIT:  I like the use of an exception here.  Because this is something that really shouldn't happen, when it does manage to get through, we catch it and throw a prompt.  This one I am happy with.

File: pg13/presentation/CryptogramEditWidget.java (hunk 4 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:117]: Totally my bad. I must have done used the autoformat funciton of eclipse. I'll change that ASAP

**Comment by rikimbo** [comment]
Regarding the UI elements having an xxxManager for all the components it needs:

Most of the UI formatting is handled by SWT and the Layout objects.  The only component where that is done manually is in the SolveWidget (or whatever I called it) to organize and display the LetterWidgets.  That part interacts directly with LayoutData objects and Widgets and stuff, so as much as I'd like to see that process move to the business layer, I think it might be unfeasable or end up looking really messy.

If you find any other logic in the presentation layer that you can move to the business layer or make a manager for, go ahead and give it a shot.  I like that you've moved the text validation to the business layer.  Check out the LetterWidget, there's validation for the text field there too, but it's more restrictive.  What's entered there needs to be a lowercase or uppercase letter (isLetter won't work since it accept accents and stuff).  Additionally, that listenner intercepts the event and sets the text to lowercase.

#############
I'm not sure how to reply to this comment so I'll just extend it.

I don't think it makes sense to move layout logic to the business layer. Since that logic is highly coupled (right word) to the UI. I was thinking of moving other logic, similar to the validation, to the business layer. I'd like to hear from Will and Lauren too just to make sure everyone is on board with my interpretation of how the business layer should look (from a code perspective) and behave.

#############
(Eric speaking again)
I agree with what you said.  Let's leave the logic involved in the arrangement and layout of gui elements with the presentation layer, but move any validation of input or objects to the business layer.

File: pg13/presentation/CryptogramEditWidget.java (hunk 4 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:117]: K, I'm gonna make an eclipse formatting thing. I'm looking at the coding standards and don't see anything about braces on the different lines. Should I add that to the coding standards?

File: pg13/presentation/CryptogramEditWidget.java (hunk 2 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:82]: Oh okay.  I looked up online how to write one of those listenners, and it told me to include that, so that's why it's there.  If it's not necessary, it's fine to take it out.

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:73]: I just feel funny about the possibility of throwing an exception every time the listenner catches a character and passes it to validate.  The user will never see them, but the fact that we expect it to happen so commonly here makes me a little bit uneasy about it.

File: pg13/presentation/CryptogramEditWidget.java (hunk 2 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:82]: TBH, I'm not sure. I almost literally just copy-pasted the code from the GUI file to the business file.

File: pg13/presentation/CryptogramEditWidget.java (hunk 4 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:117]: There's a few other places where changes are marked for bracket changes, this is the only one I commented on though.

File: pg13/business/CryptogramManager.java (hunk 1 above)
Comment (paymog) [inline pg13/business/CryptogramManager.java:41]: Well, in WPF it's not done like that at all. However, WPF is an entirely different environment. Let me read my exceptions guideline and see if I can come up with a more definitive answer :p

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:73]: I agree. Honestly, I'm not sure if us preventing them from entering a character is a good idea. Purely from an HCI perspective, if I type a character, I expect that character to become visible on the screen. Maybe it's better to let them enter anything they want and then tell them they have an invalid phrase when the press the generate preview? Thoughts?

File: pg13/presentation/CryptogramEditWidget.java (hunk 2 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:82]: Yeah, you do all the checking of the text that I was doing, just not the checking of the keycode.  When I put it in, it didn't feel like it should be there because I thought the verify listenner was just going to verify the text entered, and not mess around with keycodes in the first place.  Again, I just put the keycode checking because it  was in there on the solution I found for that on the internet.  I'm not on my home machine, otherwise I'd test it in a windows environment to make sure it works for that too.

File: pg13/presentation/CryptogramEditWidget.java (hunk 4 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:117]: Oh, I thought it was one of the things that was brought up when we were talking about standards.  Sure, add it to the document, I think we've been doing that so far.  Well, speaking for myself, I guess.

File: tests/business/CryptogramManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	private void setAndValidateValidPlainText(String plainText) {
+		this.cm.setPlaintext(plainText);
+		assertEquals(plainText, this.cm.getPlaintext());
+	}
+	
+	/**
+	 * @author PaymahnMoghadasian
+	 * @date June 1 2013
+	 * 
+	 * I'm not quite sure what constitutes invalid plaintext for a cryptogram. Somemone should extend these tests.
```
Comment (paymog) [inline tests/business/CryptogramManagerTest.java]: Someone who knows what constitutes an invalid cryptogram should either fill in this method or give me examples so I can do it

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:73]: Well, the reason I did it that was was because for a field where you're expected to enter letters or numbers (ie for a product key in an installer, for example), all it does it limit your input that way too.  For the plaintext window, really all it disallows are accented letters, and non-punctuation symbols like @ or ~ that shouldn't show up in text anyway.  We could do the validation at the point when we hit preview, but my feeling with that kind of thing has always been "Grr! Then why did you let me enter it in the first place?".  That's my preference, though.  Really it's just tomato tomahto.

**Comment by rikimbo** [comment]
Why did Travis fail on this?  I thought we had the little guy working.

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (paymog) [inline pg13/presentation/CryptogramEditWidget.java:73]: Good point. It doesn't matter for me either way. However, the question of exception vs boolean is still on the table.

File: tests/business/CryptogramManagerTest.java (hunk 6 above)
Comment (rikimbo) [inline tests/business/CryptogramManagerTest.java]: Just throw in accented letters, and other non-letters we didn't allow, eg *, @, =, whatever.  That really is all we're disallowing in the plaintext, I think.

Comment (paymog) [inline tests/business/CryptogramManagerTest.java]: Will do

File: pg13/presentation/CryptogramEditWidget.java (hunk 3 above)
Comment (rikimbo) [inline pg13/presentation/CryptogramEditWidget.java:73]: I'm actually warming up to it, since I like the way it makes the validation functions look.  I could go either way on this.

**Comment by rikimbo** [comment]
I think we're on the right track doing what you're doing here.  Do you want this merged now, or do you want to make some changes and keep committing?  Will will have some requirements from the business layer too, when he starts working on implementing the Find Screen, probably.

Same deal for when the Save button is implemented.

############
(Paymahn speaking (writing?))
This can be merged right now. I'll create another branch to extend the functionality

File: pg13/presentation/CreateScreen.java [hunk 7]
```diff
@@ -47,6 +48,7 @@ public CreateScreen(Composite parent, int style)
 		fd_composite.top = new FormAttachment(0);
 		fd_composite.left = new FormAttachment(33);
 		cmpEditWidget.setLayoutData(fd_composite);
+		setTabList(new Control[]{cmpProperties, cmpEditWidget});
```
Comment (rikimbo) [inline pg13/presentation/CreateScreen.java:51]: Does this fix the tabbing order?

**Comment by paymog** [comment]
Don't merge this. I've made more commits and I'm not sure what would happen if this gets merged and then I sync my commits.
